### PR TITLE
Form submission validation and submitting

### PR DESCRIPTION
**Pull request by ks129**
Fields that have required validation:
- [x] TextArea
- [x] ShortText
- [x] Select
- [x] Range
- [x] Radio
- [x] Checkbox
- [ ] Code (skipped for now until proper implementation)

This PR will add:
- Frontend required validation
- Submitting to backend
- Displaying backend errors to frontend

### CODE REVIEW

File: src/components/ErrorMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+interface ErrorMessageProps {
+    show: boolean,
+    message: string
+}
+
+const styles = css`
+  color: ${colors.error};
+  font-size: 18px;
+  line-height: 15px;
+  margin: 15px 0 0;
```
Comment (HassanAbouelela) [inline src/components/ErrorMessage.tsx:14]: These should use [relative units](https://www.w3schools.com/cssref/css_units.asp). See `commonStyles.tsx` for a rough idea of how px map to rems, etc.

File: src/components/ErrorMessage.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+`;
+
+export default function ErrorMessage(props: ErrorMessageProps): JSX.Element|null {
```
Comment (HassanAbouelela) [inline src/components/ErrorMessage.tsx:17]:
```suggestion
export default function ErrorMessage(props: ErrorMessageProps): JSX.Element | null {
```

File: src/components/Question.tsx [hunk 3]
```diff
@@ -1,10 +1,12 @@
+/* eslint-disable @typescript-eslint/no-empty-function */
```
Comment (HassanAbouelela) [inline src/components/Question.tsx:1]: Why was this added? Eslint isn't flagging anything without for me. You also don't enable this rule again anywhere. Could we also try to stick these as close as possible, or even inline with the function they affect to convey intent? (See line 49 for example).

File: src/commonStyles.tsx [hunk 4]
```diff
@@ -50,11 +51,81 @@ const textInputs = css`
   border-radius: 8px;
 `;
 
+const invalidStyles = css`
+  .invalid-box {
+    -webkit-appearance: none;
+    -webkit-box-shadow: 0 0 10px ${colors.error};
+    box-shadow: 0 0 10px ${colors.error};
```
Comment (HassanAbouelela) [inline src/commonStyles.tsx:58]: See comment on relative units.

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]
Looks like a solid PR so far, well done. These are a few style nitpicks, I'll add a full review later.

File: src/components/Question.tsx (hunk 3 above)
Comment (ks129) [reply src/components/Question.tsx:1]: Oh, looks like VSCode added it.

File: src/pages/FormPage.tsx [hunk 5]
```diff
@@ -107,11 +107,12 @@ function FormPage(): JSX.Element {
     useEffect(() => {
         getForm(id).then(form => {
             setForm(form);
+            setSent(true);
```
[the remark this answers is not in the document]
Comment (ks129) [reply src/pages/FormPage.tsx]: Yes, this was meant for testing.

File: src/pages/FormPage.tsx [hunk 6]
```diff
@@ -156,33 +101,115 @@ function FormPage(): JSX.Element {
...
+                        options.forEach((val, index) => {
+                            keys.set(val, `${("000" + index).slice(-4)}. ${val}`);
+                        });
+                        const pairs = {};
+                        keys.forEach((val, key) => {
+                            pairs[key] = !!prop.props.public_state.get(val);
+                        });
+                        answers[question.id] = pairs;
+                    }
+                    break;
```
Comment (HassanAbouelela) [inline src/pages/FormPage.tsx:203]:
```suggestion
                case QuestionType.Checkbox: {
                    const parsed = new Map();

                    prop.props.public_state.forEach((value, key) => {
                        if (key !== "valid" && key !== "error") {
                            answers.set(key.slice(6), value);
                        }
                    });

                    answers[question.id] = parsed;
                    break;
                }
```

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]

File: src/components/Question.tsx [hunk 7]
```diff
@@ -73,12 +107,85 @@ class RenderedQuestion extends React.Component<QuestionProp> {
...
+                    const keys: string[] = [];
+                    options.forEach((val, index) => {
+                        keys.push(`${("000" + index).slice(-4)}. ${val}`);
+                    });
+                    if (keys.every(v => !this.props.public_state.get(v))) {
+                        this.setPublicState("error", "Field must be filled.");
+                        this.setPublicState("valid", false);
+                    } else {
+                        this.setPublicState("error", "");
+                        this.setPublicState("valid", true);
```
[the remark this answers is not in the document]
Comment (ks129) [reply src/components/Question.tsx:113]: This have to re-render component, otherwise this don't trigger error showing.

File: src/components/ErrorMessage.tsx [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    message: string
+}
+
+export default function ErrorMessage(props: ErrorMessageProps): JSX.Element | null {
+    const styles = css`
+      color: ${colors.error};
+      font-size: 1.15rem;
+      line-height: 1.1rem;
+      margin: 1rem 0 0;
+      visibilty: ${props.show ? "visible" : "hidden"}
```
Comment (HassanAbouelela) [inline src/components/ErrorMessage.tsx:16]:
```suggestion
      visibility: ${props.show ? "visible" : "hidden"};
```

File: src/components/InputTypes/index.tsx [hunk 9]
```diff
@@ -34,7 +39,7 @@ export default function create_input({ question, public_state }: QuestionProp, h
     /* eslint-disable react/react-in-jsx-scope */
     switch (question.type) {
         case QuestionType.TextArea:
-            result = <TextArea handler={handler}/>;
+            result = <TextArea handler={handler} valid={valid} onBlurHandler={blurHandler} focus_ref={focus_ref}/>;
```
Comment (HassanAbouelela) [inline src/components/InputTypes/index.tsx:42]: You use different names for the same thing `onBlurHandler` across the JSX statements in this file. Pick one and stick with it for clarity.

File: src/components/ErrorMessage.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+interface ErrorMessageProps {
+    show: boolean,
+    message: string
+}
+
+const styles = css`
+  color: ${colors.error};
+  font-size: 1.15rem;
+  line-height: 1.1rem;
+  margin: 1rem 0 0;
```
[the remark this answers is not in the document]
Comment (HassanAbouelela) [reply src/components/ErrorMessage.tsx]: This is still unresolved. This suggestion only works if you set the position, otherwise elements are still pushed down. Additionally, it might be a good idea to set a negative z-index to hide the element if it gets overlapped by something else like a menu.

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]

**Review by HassanAbouelela — APPROVED** [review]
No, no, this won't do at all. You didn't remove [this line](https://github.com/python-discord/forms-frontend/blob/6ae1bb7a5de86dca3a8452eb93f00bd5bd6f8e1b/src/pages/FormPage.tsx#L252) 😛 

Thanks for keeping up with all the changes, this LGTM.

**Review by kosayoda — APPROVED** [review]
Works fine in testing.
